File: lib/infisical.js
/**
 * Cliente Infisical para Node.js — con cache local cifrada y SSH key loader en memoria pura.
 *
 * Uso:
 *   const { get, getMany, getSSHKey } = require('./secrets');
 *
 *   const dbPass = await get('DB_LOCAL_PASS');             // path por defecto: /shared
 *   const dbPass = await get('DB_LOCAL_PASS', '/shared');  // path explícito
 *   const sshKey = await getSSHKey('VPS');                  // devuelve string crudo de la key
 *
 * Variables de entorno requeridas (en .env mínimo del servicio):
 *   INFISICAL_HOST          - https://vps-contabo.tail44c420.ts.net
 *   INFISICAL_PROJECT_ID    - a332fa07-1e9f-41eb-ad25-ba701da6c5bd
 *   INFISICAL_CLIENT_ID     - UUID del client
 *   INFISICAL_CLIENT_SECRET - secret hex
 *   INFISICAL_ENV           - prod | dev | staging (default: prod)
 *
 * Cache:
 *   Memoria por proceso (Map). TTL configurable, default 5 min.
 *   No persiste a disco (cache cifrada en disco la haremos cuando se valide piloto).
 */

const https = require('https');
const path = require('path');
const fs = require('fs');

const HOST = process.env.INFISICAL_HOST || 'https://vps-contabo.tail44c420.ts.net';
const PROJECT_ID = process.env.INFISICAL_PROJECT_ID || 'a332fa07-1e9f-41eb-ad25-ba701da6c5bd';
const CLIENT_ID = process.env.INFISICAL_CLIENT_ID;
const CLIENT_SECRET = process.env.INFISICAL_CLIENT_SECRET;
const ENV = process.env.INFISICAL_ENV || 'prod';
const CACHE_TTL_MS = parseInt(process.env.INFISICAL_CACHE_TTL_MS || '300000', 10); // 5 min

if (!CLIENT_ID || !CLIENT_SECRET) {
  // Carga desde archivo de bootstrap si las env vars no están (fallback útil en local)
  const bootstrapFile = '/home/osserver/tempoclv/.infisical_admin_bootstrap.env';
  if (fs.existsSync(bootstrapFile)) {
    const content = fs.readFileSync(bootstrapFile, 'utf8');
    const lines = content.split('\n');
    for (const line of lines) {
      const m = line.match(/^INFISICAL_(\w+)=(.+)$/);
      if (m) process.env[`INFISICAL_${m[1]}`] = m[2];
    }
  }
}

// Cache: Map<cacheKey, {value, expiresAt}>
const _cache = new Map();
let _accessToken = null;
let _tokenExpiresAt = 0;

function _httpRequest(method, urlPath, body, headers = {}) {
  return new Promise((resolve, reject) => {
    const url = new URL(HOST + urlPath);
    const opts = {
      method,
      hostname: url.hostname,
      port: url.port || 443,
      path: url.pathname + url.search,
      headers: { 'Content-Type': 'application/json', ...headers },
    };
    const req = https.request(opts, (res) => {
      let data = '';
      res.on('data', (c) => (data += c));
      res.on('end', () => {
        try {
          const parsed = data ? JSON.parse(data) : null;
          resolve({ status: res.statusCode, body: parsed });
        } catch (e) {
          resolve({ status: res.statusCode, body: data });
        }
      });
    });
    req.on('error', reject);
    if (body) req.write(JSON.stringify(body));
    req.end();
  });
}

async function _ensureToken() {
  const now = Date.now();
  if (_accessToken && now < _tokenExpiresAt - 30000) return _accessToken; // 30s margin

  const clientId = process.env.INFISICAL_CLIENT_ID;
  const clientSecret = process.env.INFISICAL_CLIENT_SECRET;
  if (!clientId || !clientSecret) {
    throw new Error('INFISICAL_CLIENT_ID y INFISICAL_CLIENT_SECRET requeridos');
  }
  const { status, body } = await _httpRequest(
    'POST',
    '/api/v1/auth/universal-auth/login',
    { clientId, clientSecret }
  );
  if (status !== 200 || !body.accessToken) {
    throw new Error(`Infisical login failed: HTTP ${status} ${JSON.stringify(body).slice(0, 200)}`);
  }
  _accessToken = body.accessToken;
  _tokenExpiresAt = now + (body.expiresIn || 2592000) * 1000;
  return _accessToken;
}

/**
 * Obtiene un secret de Infisical (con cache).
 *
 * @param {string} key - nombre del secret (ej: 'DB_LOCAL_PASS')
 * @param {string} folder - folder/path (ej: '/shared', default '/shared')
 * @param {object} opts - { ttlMs, env, force }
 * @returns {Promise<string>} valor del secret
 */
async function get(key, folder = '/shared', opts = {}) {
  const env = opts.env || ENV;
  const cacheKey = `${env}:${folder}:${key}`;
  const now = Date.now();

  if (!opts.force) {
    const cached = _cache.get(cacheKey);
    if (cached && cached.expiresAt > now) return cached.value;
  }

  const token = await _ensureToken();
  const params = new URLSearchParams({
    workspaceId: PROJECT_ID,
    environment: env,
    secretPath: folder,
  });
  const { status, body } = await _httpRequest(
    'GET',
    `/api/v3/secrets/raw/${encodeURIComponent(key)}?${params}`,
    null,
    { Authorization: `Bearer ${token}` }
  );
  if (status !== 200) {
    throw new Error(`Infisical GET ${folder}/${key} failed: HTTP ${status} ${JSON.stringify(body).slice(0, 200)}`);
  }
  const value = body?.secret?.secretValue ?? body?.secretValue;
  if (value === undefined) {
    throw new Error(`Infisical: secret '${folder}/${key}' sin valor en respuesta`);
  }
  _cache.set(cacheKey, { value, expiresAt: now + (opts.ttlMs || CACHE_TTL_MS) });
  return value;
}

/**
 * Obtiene múltiples secrets de UN folder de una sola vez (más eficiente).
 *
 * @param {string} folder - path del folder (ej: '/shared')
 * @returns {Promise<Object>} { KEY1: value1, KEY2: value2, ... }
 */
async function getMany(folder = '/shared', opts = {}) {
  const env = opts.env || ENV;
  const token = await _ensureToken();
  const params = new URLSearchParams({
    workspaceId: PROJECT_ID,
    environment: env,
    secretPath: folder,
  });
  const { status, body } = await _httpRequest(
    'GET',
    `/api/v3/secrets/raw?${params}`,
    null,
    { Authorization: `Bearer ${token}` }
  );
  if (status !== 200) {
    throw new Error(`Infisical getMany ${folder} failed: HTTP ${status}`);
  }
  const result = {};
  const now = Date.now();
  for (const s of body.secrets || []) {
    result[s.secretKey] = s.secretValue;
    _cache.set(`${env}:${folder}:${s.secretKey}`, {
      value: s.secretValue,
      expiresAt: now + (opts.ttlMs || CACHE_TTL_MS),
    });
  }
  return result;
}

/**
 * Obtiene una SSH key privada de Infisical como string (para uso en memoria pura).
 * No escribe a disco.
 *
 * @param {string} which - 'VPS' o 'HOSTINGER'
 * @returns {Promise<string>} key privada como string (con BEGIN/END marcadores)
 */
async function getSSHKey(which) {
  const keyMap = {
    VPS: 'SSH_KEY_VPS_PRIVATE_ED25519',
    HOSTINGER: 'SSH_KEY_HOSTINGER_PRIVATE',
  };
  const secretName = keyMap[which.toUpperCase()];
  if (!secretName) throw new Error(`SSH key '${which}' desconocida (válidas: VPS, HOSTINGER)`);
  return get(secretName, '/ssh-keys-privadas');
}

/** Limpia el cache (útil para tests o post-rotación). */
function clearCache() {
  _cache.clear();
  _accessToken = null;
  _tokenExpiresAt = 0;
}

module.exports = { get, getMany, getSSHKey, clearCache };

File: scripts/lib/infisical.py
"""
Cliente Infisical para Python — con cache en memoria y SSH key loader sin tocar disco.

Uso:
    from lib.secrets import get, get_many, get_ssh_key_object

    db_pass = get('DB_LOCAL_PASS')                  # path default: /shared
    db_pass = get('DB_LOCAL_PASS', folder='/shared')

    # SSH key como objeto paramiko, en memoria pura
    pkey = get_ssh_key_object('VPS')
    with SSHTunnelForwarder((host, 22), ssh_pkey=pkey, ...):
        ...

Variables de entorno requeridas:
    INFISICAL_HOST          - https://vps-contabo.tail44c420.ts.net
    INFISICAL_PROJECT_ID    - a332fa07-1e9f-41eb-ad25-ba701da6c5bd
    INFISICAL_CLIENT_ID     - UUID del client (id de Universal Auth)
    INFISICAL_CLIENT_SECRET - secret hex
    INFISICAL_ENV           - prod | dev | staging (default: prod)
"""
from __future__ import annotations

import json
import os
import time
import urllib.parse
import urllib.request
import urllib.error
from io import StringIO
from typing import Any

# ── Config ───────────────────────────────────────────────────────────────────

HOST = os.getenv('INFISICAL_HOST', 'https://vps-contabo.tail44c420.ts.net')
PROJECT_ID = os.getenv(
    'INFISICAL_PROJECT_ID', 'a332fa07-1e9f-41eb-ad25-ba701da6c5bd'
)
ENV = os.getenv('INFISICAL_ENV', 'prod')
CACHE_TTL_S = int(os.getenv('INFISICAL_CACHE_TTL_S', '300'))  # 5 min

# Fallback: si las env vars no están, leer del archivo de bootstrap (útil en local)
if not os.getenv('INFISICAL_CLIENT_ID'):
    _bootstrap = '/home/osserver/tempoclv/.infisical_admin_bootstrap.env'
    if os.path.exists(_bootstrap):
        with open(_bootstrap) as f:
            for line in f:
                if line.startswith('INFISICAL_') and '=' in line:
                    k, v = line.strip().split('=', 1)
                    os.environ.setdefault(k, v)


# ── Cache + token ────────────────────────────────────────────────────────────

_cache: dict[str, tuple[Any, float]] = {}  # cacheKey -> (value, expiresAt)
_access_token: str | None = None
_token_expires_at: float = 0.0


def _http(method: str, path: str, body: Any = None, headers: dict | None = None) -> tuple[int, Any]:
    """HTTP request usando urllib (sin dependencias externas)."""
    url = HOST + path
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(url, data=data, method=method)
    req.add_header('Content-Type', 'application/json')
    for k, v in (headers or {}).items():
        req.add_header(k, v)
    try:
        with urllib.request.urlopen(req, timeout=30) as r:
            payload = r.read()
            return r.status, json.loads(payload) if payload else None
    except urllib.error.HTTPError as e:
        payload = e.read()
        try:
            return e.code, json.loads(payload)
        except Exception:
            return e.code, payload.decode(errors='ignore')


def _ensure_token() -> str:
    """Obtiene access token (login con Universal Auth) si no hay uno válido en cache."""
    global _access_token, _token_expires_at
    now = time.time()
    if _access_token and now < _token_expires_at - 30:
        return _access_token

    client_id = os.environ.get('INFISICAL_CLIENT_ID')
    client_secret = os.environ.get('INFISICAL_CLIENT_SECRET')
    if not client_id or not client_secret:
        raise RuntimeError(
            'INFISICAL_CLIENT_ID y INFISICAL_CLIENT_SECRET son requeridos'
        )

    status, body = _http(
        'POST',
        '/api/v1/auth/universal-auth/login',
        {'clientId': client_id, 'clientSecret': client_secret},
    )
    if status != 200 or not isinstance(body, dict) or 'accessToken' not in body:
        raise RuntimeError(f'Infisical login failed: HTTP {status} {str(body)[:200]}')

    _access_token = body['accessToken']
    _token_expires_at = now + body.get('expiresIn', 2_592_000)
    return _access_token


# ── API pública ──────────────────────────────────────────────────────────────


def get(key: str, folder: str = '/shared', *, env: str | None = None,
        ttl_s: int | None = None, force: bool = False) -> str:
    """Obtiene un secret de Infisical (con cache).

    Args:
        key:    nombre del secret (ej: 'DB_LOCAL_PASS')
        folder: path del folder (default '/shared')
        env:    environment (default desde INFISICAL_ENV)
        ttl_s:  TTL del cache en segundos (default CACHE_TTL_S)
        force:  si True, ignora cache y refetcha
    """
    env = env or ENV
    cache_key = f'{env}:{folder}:{key}'
    now = time.time()

    if not force and cache_key in _cache:
        value, expires_at = _cache[cache_key]
        if now < expires_at:
            return value

    token = _ensure_token()
    qs = urllib.parse.urlencode(
        {'workspaceId': PROJECT_ID, 'environment': env, 'secretPath': folder}
    )
    status, body = _http(
        'GET',
        f'/api/v3/secrets/raw/{urllib.parse.quote(key)}?{qs}',
        headers={'Authorization': f'Bearer {token}'},
    )
    if status != 200:
        raise RuntimeError(
            f"Infisical GET {folder}/{key} failed: HTTP {status} {str(body)[:200]}"
        )
    secret = (body or {}).get('secret') or body
    value = secret.get('secretValue') if isinstance(secret, dict) else None
    if value is None:
        raise RuntimeError(f"Infisical: secret '{folder}/{key}' sin valor")

    _cache[cache_key] = (value, now + (ttl_s or CACHE_TTL_S))
    return value


def get_many(folder: str = '/shared', *, env: str | None = None,
             ttl_s: int | None = None) -> dict[str, str]:
    """Obtiene TODOS los secrets de un folder de una sola llamada (más eficiente)."""
    env = env or ENV
    token = _ensure_token()
    qs = urllib.parse.urlencode(
        {'workspaceId': PROJECT_ID, 'environment': env, 'secretPath': folder}
    )
    status, body = _http(
        'GET', f'/api/v3/secrets/raw?{qs}',
        headers={'Authorization': f'Bearer {token}'},
    )
    if status != 200:
        raise RuntimeError(f'Infisical getMany {folder} failed: HTTP {status}')

    result: dict[str, str] = {}
    now = time.time()
    ttl = ttl_s or CACHE_TTL_S
    for s in (body or {}).get('secrets', []):
        k = s.get('secretKey')
        v = s.get('secretValue')
        if k is not None and v is not None:
            result[k] = v
            _cache[f'{env}:{folder}:{k}'] = (v, now + ttl)
    return result


def get_ssh_key_string(which: str) -> str:
    """Devuelve la SSH key privada como string (con BEGIN/END markers).

    Args:
        which: 'VPS' o 'HOSTINGER'
    """
    key_map = {
        'VPS': 'SSH_KEY_VPS_PRIVATE_ED25519',
        'HOSTINGER': 'SSH_KEY_HOSTINGER_PRIVATE',
    }
    secret_name = key_map.get(which.upper())
    if not secret_name:
        raise ValueError(f"SSH key '{which}' desconocida (válidas: VPS, HOSTINGER)")
    return get(secret_name, '/ssh-keys-privadas')


def get_ssh_key_object(which: str):
    """Devuelve la SSH key como objeto paramiko (en memoria, sin tocar disco).

    Uso con sshtunnel:
        from sshtunnel import SSHTunnelForwarder
        pkey = get_ssh_key_object('VPS')
        with SSHTunnelForwarder((host, 22), ssh_pkey=pkey, ...):
            ...

    Args:
        which: 'VPS' o 'HOSTINGER'
    """
    from paramiko import Ed25519Key, RSAKey
    key_str = get_ssh_key_string(which)
    buf = StringIO(key_str)
    # Ed25519 es nuestro caso. Si en el futuro hay RSA, podemos detectar por header.
    try:
        return Ed25519Key.from_private_key(buf)
    except Exception:
        buf.seek(0)
        return RSAKey.from_private_key(buf)


def clear_cache() -> None:
    """Limpia el cache (útil para tests o post-rotación)."""
    global _access_token, _token_expires_at
    _cache.clear()
    _access_token = None
    _token_expires_at = 0.0
